poutyne/framework/callbacks/tracker.py

```diff
@@ -0,0 +1,176 @@
+        # pylint: disable=unused-argument
+        named_parameters = ((n, p) for n, p in self.model.network.named_parameters() if self._keep_layer(p, n))
+        self.tracker.batch_statistic_upgrade(named_parameters)
+
+    def on_train_begin(self, logs: Dict):
+        for layer_name, layer_params in self.model.network.named_parameters():
+            self._get_layers_to_track(layer_name, layer_params)
+        self.tracker = WeightsGradientsStatsTracker(self.number_layers)
+
+    def on_train_end(self, logs: Dict):
+        self.writer.close()
+
+    def on_epoch_end(self, epoch: int, logs: Dict):
+        self._on_epoch_end_log(epoch, logs)
+
+    def _on_epoch_end_log(self, epoch: int, logs: Dict):
+        """
+        The method to define the behavior of the logging tracker into a writer.
+
+        Args:
+            epoch (int): The epoch number.
+            logs (Dict): The epoch logs dictionary.
+        """
+        pass
+
+    def _get_layers_to_track(self, layer_name: str, layer_params: torch.nn.parameter.Parameter):
+        if self._keep_layer(layer_params, layer_name):
+            self.layer_names.append(layer_name)
+
+        self.number_layers = len(self.layer_names)
+
+    def _keep_layer(self, layer_params: torch.nn.parameter.Parameter, layer_name: str):
+        layer_require_grad = layer_params.requires_grad
+        if self.keep_bias:
+            return layer_require_grad
+        return layer_require_grad and ("bias" not in layer_name)
+
+
+class WeightsGradientsStatsTracker:
+    # pylint: disable=line-too-long
+    """
+    The weights' gradient statistic tracker will compute the running absolute mean, running variance, min and max per layer.
+    The tracker is using the `Welford's online algorithm <https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm>`_
+    to compute the running absolute mean and running variance of the weights' gradients.
+    """
+
+    def __init__(self, number_layers) -> None:
+        self.number_layers = number_layers
+
+        self.reset()
+
+    def batch_statistic_upgrade(self, named_parameters: Tuple[str, torch.nn.parameter.Parameter]) -> None:
+        """
+        Accumulate the running mean, running variance, min and the max for all the layers.
+        Args:
+             named_parameters (Tuple[str, ~torch.nn.parameter.Parameter): The named parameters of the model to
+             track.
+        """
+        batch_layer_means = []
+        batch_layer_min = []
+        batch_layer_max = []
+
+        for _, layer_params in named_parameters:
+            layer_gradient = layer_params.grad
+
+            abs_value_layer_gradient = layer_gradient.abs()
+
+            batch_layer_means.append(abs_value_layer_gradient.mean().cpu().detach().numpy())
+            batch_layer_min.append(layer_gradient.min().cpu().detach().numpy())
+            batch_layer_max.append(layer_gradient.max().cpu().detach().numpy())
```

Comment:
Why does we do the absoiute mean and variance but not the absolute min and max?

Reply:
Good point. Maybe we should do both ?

Reply:
cdfc650